Trippy/Handlers/CacheDataHandler.m

```diff
@@ -159,6 +157,9 @@ - (void) fetchSavedItineraries {
                                 [[CoreDataHandler shared] saveItinerary:itinerary]; // save to local cache
                             }
                             [strongSelf.delegate addFetchedItinerary:itinerary];
+                            if ([obj isEqual:[objects lastObject]]) {
+                                [strongSelf.delegate didAddAll];
+                            }
```

Comment:
I worry about this logic. You have a for loop that enqueues a bunch of async operations. What happens if they don't process in order? 

I would suggest instead of comparing against the last object, you should actually count them as you process them. Add an atomic int property, set it to `objects.count`, and decrement as you process; and when it hits 0 again, then you can call `didAddAll`. The property being atomic should make it threadsafe and inherently synchronized.